h/browser/chrome/lib/browser-action.js

```diff
@@ -36,23 +36,88 @@
+     */
+    function setTitleIfNoBadge(tabId, title) {
+      chromeBrowserAction.getBadgeText({tabId: tabId}, function(text) {
+        if (!text) {
+          chromeBrowserAction.setTitle({tabId: tabId, title: title});
+        }
+      });
+    }
+
+    /**
+     * Set the "title" (tooltip) and badge text of the browser action.
+     * @param {integer} tabId The id of the tab to set the badge title for.
+     * @param {string} title The text to show in the tooltip.
+     * @param {string} badgeText The text to show on the badge.
+     */
+    function setTitleAndBadgeText(tabId, title, badgeText) {
+      chromeBrowserAction.setTitle({tabId: tabId, title: title});
+      chromeBrowserAction.setBadgeText({tabId: tabId, text: badgeText});
+    }
+
     /* Sets the active browser action appearance for the provided tab id. */
-    this.activate = function (tabId) {
+    this.activate = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.ACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis is active')});
+      setTitleIfNoBadge(tabId, _('Hypothesis is active'));
     };
 
     /* Sets the inactive browser action appearance for the provided tab id. */
-    this.deactivate = function (tabId) {
+    this.deactivate = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.INACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis is inactive')});
+      setTitleIfNoBadge(tabId, _('Hypothesis is inactive'));
     };
 
     /* Sets the errored browser action appearance for the provided tab id. */
-    this.error = function (tabId) {
+    this.error = function(tabId) {
       chromeBrowserAction.setIcon({tabId: tabId, path: icons[states.INACTIVE]});
-      chromeBrowserAction.setTitle({tabId: tabId, title: _('Hypothesis has failed to load')});
-      chromeBrowserAction.setBadgeText({tabId: tabId, text: '!'});
+      setTitleAndBadgeText(tabId,  _('Hypothesis has failed to load'), '!');
+    };
+
+    /**
+     * Show the number of annotations of the current page in the badge.
+     * @method
+     * @param {integer} tabId The id of the current tab.
+     * @param {string} tabUrl The URL of the current tab.
+     * @param {string} serviceUrl The URL of the Hypothesis API.
+     */
+    this.updateBadge = function(tabId, tabUrl, serviceUrl) {
+      // Fetch the number of annotations of the current page from the server,
+      // and display it as a badge on the browser action button.
+      var xhr = new XMLHttpRequest();
+      xhr.onload = function() {
+        var total = 0;
+        total = JSON.parse(this.response).total;
+        if (total > 0) {
+          var totalString = total.toString();
+          if (total > 999) {
+            totalString = '999+';
+          }
+          chromeBrowserAction.getBadgeText({tabId: tabId}, function(text) {
+            if (!text) {
```

Comment:
You've noted in the PR that this condition deals with the possibility that there's an error badge ("!") on the extension. I'd add a comment here to that effect, because it's not obvious, and I could easily see a future reorganisation losing this behaviour.